File: cspro-dev/cspro/CSEntryDroid/app/src/main/java/gov/census/cspro/engine/functions/GPSFunction.java
package gov.census.cspro.engine.functions;

import android.app.Activity;
import android.content.Intent;

import androidx.fragment.app.DialogFragment;
import androidx.appcompat.app.AppCompatActivity;


import gov.census.cspro.engine.BaseMapSelection;
import gov.census.cspro.engine.Messenger;
import gov.census.cspro.engine.functions.fragments.EngineGpsDialogFragment;
import gov.census.cspro.location.GpsReader;
import gov.census.cspro.location.LocationActivity;

public class GPSFunction implements EngineFunction {

	public static final int GPS_OPEN = 1;
	public static final int GPS_CLOSE = 2;
	public static final int GPS_READ = 3;
	public static final int GPS_READLAST = 4;
	public static final int GPS_READINTERACTIVE = 5;
	public static final int GPS_SELECT = 6;

	public static final int DIALOG_TYPE_INTERACTIVE = 0;
    public static final int DIALOG_TYPE_SELECT = 1;

	private static final String stringTrue = "1";
	private static final String stringFalse = "0";

	private int m_command;
	private int m_waitTime;
	private int m_desiredAccuracy;
	private String m_dialogText;
    private BaseMapSelection m_baseMapSelection;

	static private GpsReader m_gpsReader = new GpsReader();

	public GPSFunction(int command, int waitTime, int desiredAccuracy, String dialogText, BaseMapSelection baseMapSelection)
	{
		m_command = command;
		m_waitTime = waitTime;
		m_desiredAccuracy = desiredAccuracy;
		m_dialogText = dialogText;
        m_baseMapSelection = baseMapSelection;
    }

	public static GpsReader getReader()
	{
		return m_gpsReader;
	}

	public void runEngineFunction(Activity activity) {
        switch (m_command) {
            case GPS_OPEN:
                if (m_gpsReader.isRunning()) {
                    // Already running
                    endFunction(stringTrue);
                } else {
                    // Attempt to start
                    m_gpsReader.start(activity, new GpsReader.EnableListener() {
                        @Override
                        public void onSuccess() {
                            endFunction(stringTrue);
                        }

                        @Override
                        public void onFailure() {
                            endFunction(stringFalse);
                        }
                    });
                }
                break;

            case GPS_CLOSE:
                close();
                endFunction(stringTrue);
                break;

            case GPS_READLAST:
                endFunction(m_gpsReader.readLast());
                break;

            case GPS_READ:
				if (!m_gpsReader.isRunning()) {
					endFunction("");
				} else {
					DialogFragment dlgFragment = EngineGpsDialogFragment.newInstance(m_waitTime, m_desiredAccuracy, m_dialogText);
					dlgFragment.show(((AppCompatActivity) activity).getSupportFragmentManager(), "ModalDialogFragment");
				}
                break;

            case GPS_READINTERACTIVE:
                showInteractiveLocationDialog(activity, DIALOG_TYPE_INTERACTIVE);
                break;

            case GPS_SELECT:
                showInteractiveLocationDialog(activity, DIALOG_TYPE_SELECT);
                break;
        }
    }

    private void showInteractiveLocationDialog(Activity activity, int dialogType) {
        Intent intentSelect = new Intent(activity, LocationActivity.class);
        intentSelect.putExtra("WAIT_TIME", m_waitTime);
        intentSelect.putExtra("ACCURACY", m_desiredAccuracy);
        intentSelect.putExtra("DIALOG_TEXT", m_dialogText);
        intentSelect.putExtra("DIALOG_TYPE", dialogType);
        intentSelect.putExtra("BASE_MAP_TYPE", m_baseMapSelection.getType());
        intentSelect.putExtra("BASE_MAP_FILENAME", m_baseMapSelection.getFilename());
        activity.startActivity(intentSelect);
    }

	public static void close()
	{
		m_gpsReader.stop();
	}

	private void endFunction(String retVal)
	{
		Messenger.getInstance().engineFunctionComplete(retVal);
	}
}

File: src/wasmJsMain/kotlin/gov/census/cspro/engine/functions/GPSFunction.kt
package gov.census.cspro.engine.functions

import gov.census.cspro.location.GpsEnableListener
import gov.census.cspro.location.GpsReader
import gov.census.cspro.location.GpsReading
import gov.census.cspro.maps.BaseMapSelection
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.delay
import kotlin.js.JsAny

/**
 * Top-level JS functions for GPS
 */
@JsFun("() => { return Date.now(); }")
private external fun dateNow(): Double

/**
 * GPS Engine Function
 * Mirrors: CSEntryDroid/app/src/main/java/gov/census/cspro/engine/functions/GPSFunction.java
 */
class GPSFunction(
    private val command: Int,
    private val waitTime: Int = 0,
    private val desiredAccuracy: Int = 0,
    private val dialogText: String = "",
    private val baseMapSelection: BaseMapSelection = BaseMapSelection()
) {
    companion object {
        const val GPS_OPEN = 1
        const val GPS_CLOSE = 2
        const val GPS_READ = 3
        const val GPS_READLAST = 4
        const val GPS_READINTERACTIVE = 5
        const val GPS_SELECT = 6
        
        const val DIALOG_TYPE_INTERACTIVE = 0
        const val DIALOG_TYPE_SELECT = 1
        
        private val gpsReader = GpsReader()
        
        /**
         * Get the shared GPS reader instance
         */
        fun getReader(): GpsReader = gpsReader
    }
    
    /**
     * Execute the GPS function
     * @return The result string (GPS reading, "1" for success, "0" for failure, or "" for no reading)
     */
    suspend fun execute(): String {
        return when (command) {
            GPS_OPEN -> open()
            GPS_CLOSE -> close()
            GPS_READ -> read()
            GPS_READLAST -> readLast()
            GPS_READINTERACTIVE -> readInteractive()
            GPS_SELECT -> select()
            else -> ""
        }
    }
    
    private suspend fun open(): String {
        if (gpsReader.isRunning()) {
            return "1"
        }
        
        val result = CompletableDeferred<Boolean>()
        
        gpsReader.start(object : GpsEnableListener {
            override fun onSuccess() {
                result.complete(true)
            }
            
            override fun onFailure() {
                result.complete(false)
            }
        })
        
        return if (result.await()) "1" else "0"
    }
    
    private fun close(): String {
        gpsReader.stop()
        return "1"
    }
    
    private fun readLast(): String {
        return gpsReader.readLast()
    }
    
    private suspend fun read(): String {
        if (!gpsReader.isRunning()) {
            return ""
        }
        
        // Wait for GPS reading with timeout
        val startTime = currentTimeMillis()
        val timeoutMs = if (waitTime > 0) waitTime * 1000L else 60000L
        
        while (currentTimeMillis() - startTime < timeoutMs) {
            if (gpsReader.hasNewGPSReading(desiredAccuracy)) {
                return gpsReader.readMostAccurateGPS()
            }
            delay(200)
        }
        
        // Timeout - return best reading we have
        return gpsReader.readMostAccurateGPS()
    }
    
    private suspend fun readInteractive(): String {
        // On web, interactive mode shows a dialog with map (if available)
        // For now, fall back to regular read
        return read()
    }
    
    private suspend fun select(): String {
        // Select mode allows user to pick location on map
        // For now, fall back to regular read
        return read()
    }
    
    private fun currentTimeMillis(): Long {
        return dateNow().toLong()
    }
}

/**
 * GPS utilities for location string parsing
 * Mirrors: CSEntryDroid/app/src/main/java/gov/census/cspro/location/LocationUtils.java
 */
object LocationUtils {
    /**
     * Parse GPS string to individual components
     */
    fun parseGpsString(gpsString: String): GpsReading? {
        return GpsReading.fromGpsString(gpsString)
    }
    
    /**
     * Get latitude from GPS string
     */
    fun getLatitude(gpsString: String): Double {
        return parseGpsString(gpsString)?.latitude ?: Double.NaN
    }
    
    /**
     * Get longitude from GPS string
     */
    fun getLongitude(gpsString: String): Double {
        return parseGpsString(gpsString)?.longitude ?: Double.NaN
    }
    
    /**
     * Get altitude from GPS string
     */
    fun getAltitude(gpsString: String): Double {
        return parseGpsString(gpsString)?.altitude ?: Double.NaN
    }
    
    /**
     * Get accuracy from GPS string
     */
    fun getAccuracy(gpsString: String): Float {
        return parseGpsString(gpsString)?.accuracy ?: Float.NaN
    }
    
    /**
     * Get satellites from GPS string
     */
    fun getSatellites(gpsString: String): Int {
        return parseGpsString(gpsString)?.satellites ?: 0
    }
    
    /**
     * Calculate distance between two points (Haversine formula)
     */
    fun distance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val earthRadius = 6371000.0 // meters
        
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        
        val a = kotlin.math.sin(dLat / 2) * kotlin.math.sin(dLat / 2) +
                kotlin.math.cos(Math.toRadians(lat1)) * kotlin.math.cos(Math.toRadians(lat2)) *
                kotlin.math.sin(dLon / 2) * kotlin.math.sin(dLon / 2)
        
        val c = 2 * kotlin.math.atan2(kotlin.math.sqrt(a), kotlin.math.sqrt(1 - a))
        
        return earthRadius * c
    }
}

// Math utility for web
private object Math {
    fun toRadians(degrees: Double): Double = degrees * kotlin.math.PI / 180.0
}
